File: python/problems/leetcode/0005_longest_palindromic_substring/solution.py
"""LeetCode 5: Longest Palindromic Substring."""

from __future__ import annotations


class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        start = end = 0

        def expand(l: int, r: int) -> None:
            nonlocal start, end
            while l >= 0 and r < n and s[l] == s[r]:
                if r - l > end - start:
                    start, end = l, r
                l -= 1
                r += 1

        for i in range(n):
            expand(i, i)
            expand(i, i + 1)
        return s[start : end + 1]


if __name__ == "__main__":
    assert Solution().longestPalindrome("babad") in ("bab", "aba")
    assert Solution().longestPalindrome("cbbd") == "bb"
    print("0005 OK")

File: cpp/problems/leetcode/0005_longest_palindromic_substring/solution.cpp
// LeetCode 5
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int n = (int)s.size(), st = 0, en = 0;
        auto expand = [&](int l, int r) {
            while (l >= 0 && r < n && s[l] == s[r]) {
                if (r - l > en - st) st = l, en = r;
                --l, ++r;
            }
        };
        for (int i = 0; i < n; ++i) {
            expand(i, i);
            expand(i, i + 1);
        }
        return s.substr(st, en - st + 1);
    }
};

int main() {
    string a = Solution().longestPalindrome("babad");
    assert(a == "bab" || a == "aba");
    assert(Solution().longestPalindrome("cbbd") == "bb");
    cout << "0005 OK" << endl;
    return 0;
}
